### PR TITLE
Refactor tzlocal.__init__ and fix "local" returned when no zone found

### LINKED ISSUE #44: get_localzone returns 'local' on alpine linux docker container (Python:alpine)
When i run `tzlocal.get_localzone()` it returns `<StaticTzInfo 'local'>` 

Steps to recreate
1. Get into shell of container: `docker run --rm -it python:alpine /bin/sh`
2. Install tzlocal `pip install tzlocal`
3. Launch python `python`
4. Import and test
```python
>>> tzlocal.get_localzone()
<StaticTzInfo 'local'>
```

output of /etc/localtime
```bash
cat /etc/localtime 
TZif2UTCTZif2UTC
UTC0
```
`/usr/local/etc/localtime` does not exist.

### PR DESCRIPTION
**Pull request by benjaoming**
TODO:

 - [x] Fix #44

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,9 +1,11 @@
 from setuptools import setup, find_packages
 
-version = '1.5.dev0'
+
+from tzlocal import __version__
```
Comment (regebro) [inline setup.py]: Importing the local package from setup.py is a bad idea for multiple reasons.

File: tzlocal/__init__.py [hunk 2]
```diff
@@ -1,7 +1,34 @@
 import sys
-if sys.platform == 'win32':
-    from tzlocal.win32 import get_localzone, reload_localzone
-elif 'darwin' in sys.platform:
-    from tzlocal.darwin import get_localzone, reload_localzone
-else:
-    from tzlocal.unix import get_localzone, reload_localzone
+
+__version__ = "1.5dev0"
```
Comment (regebro) [inline tzlocal/__init__.py]: We can skip this a __version__ isn't required.

If we must have one, this works:

```
import pkg_resources  # part of setuptools
version = pkg_resources.require("tzlocal")[0].version
```

File: tzlocal/__init__.py [hunk 3]
```diff
@@ -1,7 +1,34 @@
...
+_cache_tz = None
+
+def get_localzone():
+    """Returns the zoneinfo-based tzinfo object that matches the local system's timezone."""
+    global _cache_tz
+
+    if _cache_tz is not None:
+        return _cache_tz
+
+    # Avoid importing sub-modules at load time, breaks running setup.py
```
Comment (regebro) [inline tzlocal/__init__.py]: Only if you import the package from setup.py, which you shouldn't do.

File: tzlocal/darwin.py [hunk 4]
```diff
@@ -27,7 +26,7 @@ def __exit__(self, type, value, traceback):
     from subprocess import Popen
 
 
-def _get_localzone(_root='/'):
+def get_localzone_name(_root='/'):
```
Comment (regebro) [inline tzlocal/darwin.py]: No, this function must return a timezone, as the configured timezone doesn't have to be a name.

File: setup.py (hunk 1 above)
Comment (benjaoming) [reply setup.py]: That's why the `__init__` now doesn't import anything at module load time... this is tested many places... makes it easier to maintain the version. The purpose is that you now have `__version__` in `__init__`

File: tzlocal/__init__.py (hunk 2 above)
Comment (benjaoming) [reply tzlocal/__init__.py]: It's a convention

File: tzlocal/__init__.py (hunk 3 above)
Comment (benjaoming) [reply tzlocal/__init__.py]: As said, importing `__version__` in `setup.py` is just a small DRY thing.

There are reasons why you shouldn't import stuff in `__init__`, in general I keep them as empty as possible. In this case, it's a matter of keeping an API.

http://foobar.lu/wp/2012/05/13/a-comprehensive-step-through-python-packaging-a-k-a-setup-scripts/

File: tzlocal/darwin.py (hunk 4 above)
Comment (benjaoming) [reply tzlocal/darwin.py]: The purpose of the refactor was to move some generalized functionality from the system specific modules to a general scope, which currently is `tzlocal.__init__`

Comment (benjaoming) [reply tzlocal/darwin.py]: If you look below, the only change is that it now doesn't call `pytz.timezone`, it seemed more uniform to the overall way that the system-specific modules behaved

Comment (regebro) [reply tzlocal/darwin.py]: Yes, but these functions must return a timezone, as the configure time zone might not have a name.

File: setup.py (hunk 1 above)
Comment (regebro) [reply setup.py]: See the comment on that code in `__init__`

File: tzlocal/__init__.py [hunk 5]
```diff
@@ -1,7 +1,35 @@
...
+def get_localzone():
+    """Returns the zoneinfo-based tzinfo object that matches the local system's timezone."""
+    global _cache_tz
+
+    if _cache_tz is not None:
+        return _cache_tz
+
+    if _cache_tz == 'local':
+        import pytz
+        raise pytz.UnknownTimeZoneError("Got timezone 'local', this is not a valid local timezone code")
```
Comment (regebro) [inline tzlocal/__init__.py]: Local is valid.

**Comment by regebro** [comment]
Thank you for your pull request, it's appreciated. However, I will close this, as most of it is you attempting to change behavior that you understood as problematic, but is the correct and expected behavior.

If you want to move the cache handling to `__init__.py` you can make a new pull request to that effect.

**Comment by benjaoming** [comment]
Nah man, I'm fine

**Comment by regebro** [comment]
OK, thanks.

**Pull request closed without merging**